Framework/Core/include/Framework/TableTreeHelpers.h

```diff
@@ -22,122 +22,372 @@ namespace o2::framework
+};
+
+template <>
+struct ROOTTypeString<uint16_t> {
+  static constexpr char const* str = "/s";
+};
+
+template <>
+struct ROOTTypeString<uint32_t> {
+  static constexpr char const* str = "/i";
+};
+
+template <>
+struct ROOTTypeString<uint64_t> {
+  static constexpr char const* str = "/l";
+};
+
+template <>
+struct ROOTTypeString<int8_t> {
+  static constexpr char const* str = "/B";
+};
+
+template <>
+struct ROOTTypeString<int16_t> {
+  static constexpr char const* str = "/S";
+};
+
+template <>
+struct ROOTTypeString<int32_t> {
+  static constexpr char const* str = "/I";
+};
+
+template <>
+struct ROOTTypeString<int64_t> {
+  static constexpr char const* str = "/L";
+};
+
+template <>
+struct ROOTTypeString<float> {
+  static constexpr char const* str = "/F";
+};
+
+template <>
+struct ROOTTypeString<double> {
+  static constexpr char const* str = "/D";
+};
+
+template <typename T>
+constexpr auto ROOTTypeString_t = ROOTTypeString<T>::str;
+
+template <typename T>
+class ColumnToBranch : public ColumnToBranchBase
 {
+ public:
+  ColumnToBranch(TTree* tree, arrow::ChunkedArray* column, arrow::Field* field, int size = 1)
+    : ColumnToBranchBase(column, field, size)
+  {
+    if constexpr (std::is_pointer_v<T>) {
+      mLeaflist = mBranchName + "[" + std::to_string(listSize) + "]" + ROOTTypeString_t<std::remove_pointer_t<T>>;
+    } else {
+      mLeaflist = mBranchName + ROOTTypeString_t<T>;
+    }
+    mBranch = tree->GetBranch(mBranchName.c_str());
+    if (mBranch == nullptr) {
+      mBranch = tree->Branch(mBranchName.c_str(), (char*)nullptr, mLeaflist.c_str());
+    }
+    if constexpr (std::is_same_v<bool, std::remove_pointer_t<T>>) {
+      mCurrent = new bool[listSize];
+      mLast = mCurrent + listSize;
+      accessChunk(0);
+    } else {
+      accessChunk(0);
+    }
+  }
 
  private:
-  TTree* mTreePtr;
+  std::string mLeaflist;
 
-  // a list of BranchIterator
-  std::vector<std::unique_ptr<BranchIterator>> mBranchIterators;
+  void resetBuffer() override
+  {
+    if constexpr (std::is_same_v<bool, std::remove_pointer_t<T>>) {
+      if (O2_BUILTIN_UNLIKELY((*mCurrentPos - mFirstIndex) * listSize >= getCurrentArray()->length())) {
+        nextChunk();
+      }
+    } else {
+      if (O2_BUILTIN_UNLIKELY((mCurrent + (*mCurrentPos - mFirstIndex) * listSize) >= mLast)) {
+        nextChunk();
+      }
+    }
+    accessChunk(*mCurrentPos);
+    mBranch->SetAddress((void*)(mCurrent + (*mCurrentPos - mFirstIndex) * listSize));
+  }
 
-  // table to convert
-  std::shared_ptr<arrow::Table> mTable;
+  auto getCurrentArray() const
+  {
+    if (listSize > 1) {
+      return std::static_pointer_cast<o2::soa::arrow_array_for_t<std::remove_pointer_t<T>>>(std::static_pointer_cast<arrow::FixedSizeListArray>(mColumn->chunk(mCurrentChunk))->values());
+    } else {
+      return std::static_pointer_cast<o2::soa::arrow_array_for_t<std::remove_pointer_t<T>>>(mColumn->chunk(mCurrentChunk));
+    }
+  }
+
+  void nextChunk() override
```

Comment:
I was actually thinking we cannot do that with the current implementation of the cursor, since it relies on the chunks in each column to have the same number of elements, no?

Reply:
This is, in its essence, what we do in ColumnIterator for the RowView. Number of elements can still be the same, even if the elements are composed of several plain values.